rfc/system/3844-artifact-service.md

```diff
@@ -0,0 +1,148 @@
+
+You should get a form in which you can just put in that string, and run with it.
+
+Type checking will happen in Admin as soon as the CreateExecutionRequest comes in, it can retrieve the underlying workflow/task output types and perform the type check.
+
+### Introduction of Artifact
+If you squint at the above from a one-thousand meter view, the behaviour of the URL starts to look like a cataloging service. There's a unique identifier for everything produced by Flyte, it should be usable across the platform and it should be persisted. Since the catalog name was already taken, we are calling this Artifact. An Artifact is nothing more than a tracked piece of metadata associated with a Literal and LiteralType.
+
+* Artifacts have a unique identifier per Flyte installation 
+* All outputs in Flyte automatically come with an Artifact (every output of every task/workflow execution).
+* Artifacts can have additional metadata data
+  * Can keep type information (dataset, model, maybe a "value" which is used for Scalars that are important)
+  * Can have aliases, so you should be able to alias an artifact with `latest` and then reference it.
+
+Data modeling side note: currently the Flyte URL produced by the PRs have two forms - the node level output, or the task execution level output. For the purpose of writing to the Artifact store, we're going to take the task output.
+
+### Artifact User Experience
+The Artifact story should support the following UX. All code examples subject to change/debate.
+
+#### Use an existing artifact locally
+Use case: I want to inspect the data locally, download it and use it.
+
+```python
+a = remote.get("flyte://blah")
+a = Artifact.get("flyte://blah", remote)
+a.download()  # errors if it's not an off-loaded type, downloads otherwise
+a.as_python_val(type_hint=...)  # call to the TypeEngine, guessing the Python type if there's no type hint
+```
+
+#### Use an existing artifact as input to a new workflow execution
+Use case: A task in another workflow produced some data. I now want to use that data as the input to a new and different workflow run.
+
+```python
+df_artifact = Artifact("flyte://a1")
+remote.execute(wf, inputs={"a": df_artifact})
+```
+
+Since the Artifact service can take a link and return a Literal, this part is pretty simple since remote.execute already can take Literals. Will just have to fetch and add it to the LiteralMap.  When remote runs, it’ll have to fetch the Literal.
+
+Note: with this construct, two different versions/aliases/etc of the same dataset/artifact can be used in one workflow - think about a workflow that compares the performance of two models.
+
+#### Specify metadata at Artifact creation time
+Use case: When this task or workflow runs, I want to tag it with certain tags and give it a version alias.
+
+```python
+@task
+def t1() -> Annotated[nn.Module, Artifact(name="my.artifact.name", alias={"version": "latest"}, type=Model)]
```

Comment:
how will `alias` be populated at runtime? will it always be `"latest"` or can it be overridden for a specific execution?  How will this work for cached workflows. 

In an ideal world, we'd want to fetch artifacts by some name + execution id.